presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/TypeMatchingChecker.java

```diff
@@ -0,0 +1,169 @@
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
+            this.types = requireNonNull(types, "types is null");
+        }
+
+        @Override
+        public Void visitAggregation(AggregationNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            AggregationNode.Step step = node.getStep();
+
+            // Do checking only when the step is SINGLE since FINAL and PARTIAL aggregation might have intermediate representation types
+            if (step.equals(AggregationNode.Step.SINGLE)) {
+                checkSignatureTypeMatching(node.getFunctions());
+                checkFunctionCallTypeMatching(node.getAggregations());
+            }
+
+            return null;
+        }
+
+        @Override
+        public Void visitWindow(WindowNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            checkSignatureTypeMatching(node.getSignatures());
+            checkFunctionCallTypeMatching(node.getWindowFunctions());
+
+            return null;
+        }
+
+        @Override
+        public Void visitProject(ProjectNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            checkExpressionTypeMatching(node.getAssignments());
+
+            return null;
+        }
+
+        @Override
+        public Void visitUnion(UnionNode node, Void context)
+        {
+            visitPlan(node, context);
+
+            checkMappingTypeMatching(node.getSymbolMapping());
+
+            return null;
+        }
+
+        private void checkMappingTypeMatching(ListMultimap<Symbol, Symbol> symbolMapping)
+        {
+            for (Symbol keySymbol : symbolMapping.keySet()) {
+                List<Symbol> valueSymbols = symbolMapping.get(keySymbol);
+                Type type = types.get(keySymbol);
+                for (Symbol valueSymbol : valueSymbols) {
+                    verifyTypeSignature(keySymbol, type.getTypeSignature(), types.get(valueSymbol).getTypeSignature());
+                }
+            }
+        }
+
+        private void checkSignatureTypeMatching(Map<Symbol, Signature> functions)
+        {
+            for (Map.Entry<Symbol, Signature> entry : functions.entrySet()) {
+                TypeSignature typeSignature = types.get(entry.getKey()).getTypeSignature();
+                TypeSignature functionTypeSignature = entry.getValue().getReturnType();
+                verifyTypeSignature(entry.getKey(), typeSignature, functionTypeSignature);
+            }
+        }
+
+        private void checkFunctionCallTypeMatching(Map<Symbol, FunctionCall> functionCalls)
+        {
+            for (Map.Entry<Symbol, FunctionCall> entry : functionCalls.entrySet()) {
+                Type type = types.get(entry.getKey());
```

Comment:
Might be clearer if you rename these to expectedType and actualType in all your check methods.